api/src/main/java/org/pmiops/workbench/institution/InstitutionServiceImpl.java

```diff
@@ -0,0 +1,146 @@
+import org.pmiops.workbench.db.model.DbInstitution;
+import org.pmiops.workbench.db.model.DbInstitutionEmailAddress;
+import org.pmiops.workbench.db.model.DbInstitutionEmailDomain;
+import org.pmiops.workbench.db.model.DbStorageEnums;
+import org.pmiops.workbench.model.Institution;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
+public class InstitutionServiceImpl implements InstitutionService {
+
+  private final Provider<InstitutionDao> institutionDaoProvider;
+  private final Provider<InstitutionEmailDomainDao> institutionEmailDomainDaoProvider;
+  private final Provider<InstitutionEmailAddressDao> institutionEmailAddressDaoProvider;
+
+  @Autowired
+  InstitutionServiceImpl(
+      Provider<InstitutionDao> institutionDaoProvider,
+      Provider<InstitutionEmailDomainDao> institutionEmailDomainDaoProvider,
+      Provider<InstitutionEmailAddressDao> institutionEmailAddressDaoProvider) {
+    this.institutionDaoProvider = institutionDaoProvider;
+    this.institutionEmailDomainDaoProvider = institutionEmailDomainDaoProvider;
+    this.institutionEmailAddressDaoProvider = institutionEmailAddressDaoProvider;
+  }
+
+  @Override
+  public List<Institution> getInstitutions() {
+    return StreamSupport.stream(institutionDaoProvider.get().findAll().spliterator(), false)
+        .map(this::toModelClass)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public Optional<Institution> getInstitution(final String id) {
+    return getDbInstitution(id).map(this::toModelClass);
+  }
+
+  @Override
+  public Institution createInstitution(final Institution institutionToCreate) {
+    return toModelClass(saveInstitution(institutionToCreate, new DbInstitution()));
+  }
+
+  @Override
+  public boolean deleteInstitution(final String id) {
+    return getDbInstitution(id)
+        .map(
+            dbInst -> {
+              institutionDaoProvider.get().delete(dbInst);
+              return true;
+            })
+        .orElse(false);
+  }
+
+  @Override
+  public Optional<Institution> updateInstitution(
+      final String id, final Institution institutionToUpdate) {
+    return getDbInstitution(id)
+        .map(dbInst -> toModelClass(saveInstitution(institutionToUpdate, dbInst)));
+  }
+
+  private Optional<DbInstitution> getDbInstitution(String id) {
+    return institutionDaoProvider.get().findOneByShortName(id);
+  }
+
+  private DbInstitution saveInstitution(final Institution modelClass, final DbInstitution dbClass) {
+    dbClass.setShortName(modelClass.getShortName());
+    dbClass.setDisplayName(modelClass.getDisplayName());
+    dbClass.setOrganizationTypeEnum(
+        DbStorageEnums.organizationTypeToStorage(modelClass.getOrganizationTypeEnum()));
+    dbClass.setOrganizationTypeOtherText(modelClass.getOrganizationTypeOtherText());
+
+    // save so the domain and address DAOs have something to reference
+    institutionDaoProvider.get().save(dbClass);
```

Comment:
```suggestion
    dbClass = institutionDaoProvider.get().save(dbClass);
```
Spring recommends that you should use the return object as it may be different altogether: https://docs.spring.io/spring-data/commons/docs/current/api/org/springframework/data/repository/CrudRepository.html#save-S-

Reply:
wow, yikes